### PR TITLE
Adds support for storing read status in meta db

### PR DESCRIPTION
**Pull request by garethbowen**
Adds a migration to extract the read status from the docs and
into the meta db.

Adds an endpoint to handle the lazy creation of the user meta
db.

Also improves the migration output to be formatted rather than
stringifying the Error.

medic/medic-webapp#2418

### CODE REVIEW

File: migrations/extract-read-status.js [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    db.request(params, callback);
+  });
+};
+
+const ensureDbExists = (username, dbName, callback) => {
+  db.db.get(dbName, err => {
+    if (err && err.statusCode === 404) {
+      return createDb(username, dbName, callback);
+    }
+    callback();
```
Comment (SCdF) [inline migrations/extract-read-status.js]: Do you want to be checking for just `err` when it's not a 404 as well?

File: db.js [hunk 2]
```diff
@@ -116,6 +117,31 @@ if (couchUrl) {
...
+      }
+      console.log(`        Processing ${skip} to ${skip + batchSize} docs of ${response.total_rows} total`);
+      iteratee(response.rows, err => {
+        const keepGoing = response.total_rows > (skip + batchSize);
+        callback(err, keepGoing);
+      });
+    });
+  };
+
+  module.exports.batch = (query, iteratee, batchSize, callback) => {
```
Comment (SCdF) [inline db.js]: It's also worth considering that if we're having a technique elevated into the "standard library" (as it were), we probably want to do skipping the more correct / performant way, using keys, as opposed to numeric record skipping.

File: controllers/create-user-db.js [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+const createDatabase = (username, callback) => {
+  const dbName = getDbName(username);
+  db.db.create(dbName, err => {
+    if (err) {
+      return callback(err);
+    }
+    const params = {
+      db: dbName,
+      path: '/_security',
+      method: 'put',
```
Comment (garethbowen) [inline controllers/create-user-db.js]: Done.

File: migrations/extract-read-status.js (hunk 1 above)
Comment (garethbowen) [inline migrations/extract-read-status.js]: Yeah. I'm not sure what that `err` would be at this stage if not a 404, but I've modified the code to pass it to the callback so at least we'll see it reported.

File: db.js (hunk 2 above)
Comment (garethbowen) [inline db.js]: Yes probably but it's complicated so would need extensive testing. I've raised a separate issue to investigate this: medic/medic-webapp#3553

File: lib/user-db.js [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+const putDdoc = (dbName, callback) => {
+  db.use(dbName).insert(ddoc, callback);
+};
+
+module.exports = {
+  getDbName: username => `medic-user-${username}-meta`,
+  create: (username, callback) => {
+    const dbName = module.exports.getDbName(username);
+    createDb(dbName, err => {
```
Comment (SCdF) [inline lib/user-db.js]: Can I getta `async.series`. Gotta kill the JS callback hate pyramid

File: db.js [hunk 5]
```diff
@@ -116,6 +117,42 @@ if (couchUrl) {
...
+        callback(err, keepGoing);
+      });
+    });
+  };
+
+  /**
+   * Run an operation over all documents returned from the query in batches.
+   *
+   * query function     Called with a skip number and a callback to
+   *                    invoke with the next batch.
```
Comment (garethbowen) [inline db.js]: Done.

File: lib/user-db.js (hunk 4 above)
Comment (garethbowen) [inline lib/user-db.js]: Here's your `async.series` and I threw in a few `async.apply` calls for free!

**Pull request merged**